### DISCUSSION TITLE
Boost error when running the software after compilation on Linux

### DISCUSSION THREAD

**Discussion by max1c** [SCIInstitute/SCIRun#2281] [General] [answered]
**Describe the bug**

After compilation seeing boost error when trying to run the application.

$ ./SCIRun
terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
  what():  boost::filesystem::status: Function not implemented: "./src"
Aborted

**To Reproduce**
1. Go to /opt/SCIRun-5.0-beta.Z/bin/SCIRun
2. ./SCIRun

**Version**

The release v5.0-beta.Z version: https://github.com/SCIInstitute/SCIRun/releases/tag/v5.0-beta.Z

**Desktop (please complete the following information):**

$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic

**Comment by dcwhite** [comment]
If you used gcc, could you run your build through the debugger? 
```
gdb ./SCIRun
run
bt
```
will generate a backtrace and I'll have more to work with. I've seen similar filesystem errors in the past, but I don't have an 18.04 machine to test on (SCIRun ought to work fine on 20.04).

**Comment by max1c** [comment]
```
$ gdb ./SCIRun 
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./SCIRun...(no debugging symbols found)...done.
(gdb) run
Starting program: /opt/SCIRun-5.0-beta.Z/bin/SCIRun/SCIRun 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
  what():  boost::filesystem::status: Function not implemented: "/opt/SCIRun-5.0-beta.Z/bin/SCIRun/src"

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00002aaabf771921 in __GI_abort () at abort.c:79
#2  0x00002aaabee7e957 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00002aaabee84ae6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00002aaabee84b21 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00002aaabee84d54 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00002aaabd5eb17e in boost::filesystem::detail::status(boost::filesystem::path const&, boost::system::error_code*) () from /opt/SCIRun-5.0-beta.Z/bin/SCIRun/lib/libCore_Util_Legacy.so
#7  0x00002aaabd5db1cb in SciEnvironmentBuilder::test_and_set_various_source_dirs(boost::filesystem::path const&, boost::filesystem::path&) () from /opt/SCIRun-5.0-beta.Z/bin/SCIRun/lib/libCore_Util_Legacy.so
#8  0x00002aaabd5d79f0 in SCIRun::create_sci_environment(char**, char const*) () from /opt/SCIRun-5.0-beta.Z/bin/SCIRun/lib/libCore_Util_Legacy.so
#9  0x000055555555d0df in mainImpl(int, char const**, char**) ()
#10 0x00002aaabf752bf7 in __libc_start_main (main=0x55555555ce90 <main>, argc=1, argv=0x7fffffffe828, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe818) at ../csu/libc-start.c:310
#11 0x000055555555cfca in _start ()
```

**Comment by max1c** [comment]
I tried building under ubuntu 20.04 but now get:
[ 74%] Completed 'Python_external'
[ 74%] Built target Python_external
make: *** [Makefile:84: all] Error 2

No other errors or anything. Is there a list of apt dependencies necessary to be installed that's more detailed than this: http://sciinstitute.github.io/SCIRun/build.html#linux ?

lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal

**Comment by dcwhite** [comment]
That backtrace helps. I think it's a bug in boost, but at least I can add some error checking on our side.

I'll need more of the build log to diagnose the failed 20.04 build. If you were running a `make -j6` for instance, try a single-core make first--sometimes the multicore make gets tripped up on the Python part of the build.

**Comment by max1c** [comment]
> That backtrace helps. I think it's a bug in boost, but at least I can add some error checking on our side.
> 
> I'll need more of the build log to diagnose the failed 20.04 build. If you were running a `make -j6` for instance, try a single-core make first--sometimes the multicore make gets tripped up on the Python part of the build.

Yes, that was indeed the issue. Here is the error after running make without -j: 
/usr/lib/qt5/bin/uic: error while loading shared libraries: libQt5Core.so.5: cannot open shared object file: No such file or directory
make[5]: *** [Interface/Modules/Base/CMakeFiles/Interface_Modules_Base.dir/build.make:62: Interface/Modules/Base/ui_ModuleDialogBasic.h] Error 127
make[4]: *** [CMakeFiles/Makefile2:2718: Interface/Modules/Base/CMakeFiles/Interface_Modules_Base.dir/all] Error 2
make[3]: *** [Makefile:152: all] Error 2
make[2]: *** [CMakeFiles/SCIRun_external.dir/build.make:126: download/FwdInvToolkit/Stamp/SCIRun_external/SCIRun_external-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:121: CMakeFiles/SCIRun_external.dir/all] Error 2
make: *** [Makefile:84: all] Error 2

Apparently a bug. Fixed with this: 
`sudo strip --remove-section=.note.ABI-tag /usr/lib/x86_64-linux-gnu/libQt5Core.so.5`

**Comment by max1c** [comment, marked as answer]
I compiled SCIRun under Ubuntu 20.04 and it's working. This can be closed. Thanks for help .
